### ISSUE TITLE
max_clusters

### ISSUE THREAD

**Issue by droctothorpe** [dask/dask-gateway#313]
Hello, @jcrist.

Unless I'm missing something (which is very possible), or it was renamed, it looks like the `max_clusters` option was removed during the re-architecture. There's no reference to it in the latest docs or source.

Are there any plans to reintroduce it?

**Comment by jcrist** [comment]
There are not any plans to reintroduce it. Things that are per-user (rather than   are hard to support in an atomic and consistent way. Take k8s for example, we store all the cluster information as custom k8s objects. Supporting an atomic "create a new cluster if it's less than a limit otherwise error" operation isn't easy to do without plugging in deeper to the k8s infrastructure (I believe we could do this with an admission controller, but it'd still be tricky). Doing it in the dask-gateway api server is prone to race conditions if running more than one instance.

If this is something that you need, you could do this with a namespace-per-user, and a `ResourceQuota` on `DaskCluster` objects (https://kubernetes.io/docs/concepts/policy/resource-quotas/). This would be a more k8s-native solution, and removes dask-gateway from having to think about these things itself. Would this solution be sufficient for your needs?

**Comment by jcrist** [comment]
Actually, if we changed the semantics from "if exceeding the limit on n-clusters, error" to "don't start a cluster if there's already n clusters, queue instead", then this could be done inside the controller (and supported by other platforms as well), but it's still an added complication I'd like to avoid if other solutions are sufficient.

**Comment by TomAugspurger** [comment]
I'm planning to explore this for pangeo's deployments (https://github.com/pangeo-data/pangeo-cloud-federation/issues/693).

The complication is that we'd like to avoid a namespace-per user. Ideally we'd be able to have a quota per label (since we label the pods with the JupyterHub username) but that may not be possible in kubernetes today (https://github.com/pangeo-data/pangeo-cloud-federation/issues/693).

**Comment by jcrist** [comment]
Hmmm, from reading that issue it sounds like relying on a namespace-per-user wouldn't work for you? If that's the case, I'd be open to going down the route described here: https://github.com/dask/dask-gateway/issues/313#issuecomment-694260346. We could then support a limit on the number of clusters per user - creating additional clusters beyond that limit would queue (but not error, erroring would be harder to support, but still doable). I'd be against extending this to resource limits per user (rather than per cluster), as that's much harder to support (still doable, but harder).

**Comment by TomAugspurger** [comment]
> it sounds like relying on a namespace-per-user wouldn't work for you?

I'm not certain, but I think that'd be challenging, given our infrastructure around deploying everything.

> creating additional clusters beyond that limit would queue (but not error, erroring would be harder to support, but still doable).

Either of these is fine by me, as long as we have some way to inform the user that they Cluster creation is queued (possibly including a note that they've exceeded their personal limit on the number of clusters). I agree that erroring is probably what would be expected by our users, but hard to say.

> I'd be against extending this to resource limits per user (rather than per cluster), as that's much harder to support (still doable, but harder).

That sounds fine to me. We can set our per-cluster resource limits in tandem with our per-user number of clusters limit.

**Comment by droctothorpe** [comment]
I agree that resource limits per user as opposed to cluster is overkill. No more than `n` clusters per user is all we need. That, in conjunction with the per-cluster resource limits forms a concrete upper limit per user anyway. I agree that it makes more sense to have the controller handle this. An informative error message would definitely be helpful.

**Comment by droctothorpe** [comment]
Since it came up, we're also evaluating the pros and cons of a namespace-per-user model. 

Please correct me if I'm wrong, but it seems like DG makes the assumption that the namespaces will be generated ahead of time. This will be difficult given that our platform is dynamic as opposed to bespoke and we anticipate hundreds of users.

If we go down the namespace-per-user model, we will need to implement some programmatic namespace creation / deletion logic to automate that step, which is doable, it will just take some dev time. 

It may be valuable to provide a turnkey or reference implementation for this if other users are inhibited from implementing a namespace-per-user model for the same reason.

If we wind up implementing it ourselves, we'll definitely try to contribute it upstream.

**Comment by jcrist** [comment]
> Please correct me if I'm wrong, but it seems like DG makes the assumption that the namespaces will be generated ahead of time. This will be difficult given that our platform is dynamic as opposed to bespoke and we anticipate hundreds of users.

This is correct. Since dask-gateway doesn't maintain a users table (a cluster has a user as metadata, but there is no table of users maintained by dask-gateway), I don't think user-management things like this belong in dask-gateway. We could add some hooks to make it work, but they might not support all things you'd like to do.

If you're using JupyterHub, I'd do this by implementing the `add_user` hook in the JupyterHub authenticator (https://github.com/jupyterhub/jupyterhub/blob/21eaf0dd9f76c9eed1757280b90e580218db9f58/jupyterhub/auth.py#L588). This gets called once for every new user (also called on hub-restart, so it would need to debounce), and could create a namespace or any other resources needed for that user. This is nice in that it happens once external to dask-gateway, so the namespace check wouldn't need to happen when creating each cluster.

To summarize, two options (as I see them):
- Implement this outside of dask-gateway with whatever user-management system you're using/building. Create the namespace (and whatever else) when a new user is created.
- Add a pre-cluster-creation hook in dask-gateway that runs your custom code whenever a new cluster is created. This would need to debounce multiple executions, as we'd have no way to know if it already ran for a user or not.

I'd prefer you use the first one if possible (less stuff for dask-gateway to do), but the latter is also fine.

**Comment by droctothorpe** [comment]
Thanks, @jcrist. Really appreciate the advice. We are planning to migrate to JH (from an in-house solution) and implemented our own auth already; the `add_user` hook sounds like a perfect fit. cc @txu0393.

**Comment by droctothorpe** [comment]
> I'm planning to explore this for pangeo's deployments ([pangeo-data/pangeo-cloud-federation#693](https://github.com/pangeo-data/pangeo-cloud-federation/issues/693)).
> 
> The complication is that we'd like to avoid a namespace-per user. Ideally we'd be able to have a quota per label (since we label the pods with the JupyterHub username) but that may not be possible in kubernetes today ([pangeo-data/pangeo-cloud-federation#693](https://github.com/pangeo-data/pangeo-cloud-federation/issues/693)).



> I'm planning to explore this for pangeo's deployments ([pangeo-data/pangeo-cloud-federation#693](https://github.com/pangeo-data/pangeo-cloud-federation/issues/693)).
> 
> The complication is that we'd like to avoid a namespace-per user. Ideally we'd be able to have a quota per label (since we label the pods with the JupyterHub username) but that may not be possible in kubernetes today ([pangeo-data/pangeo-cloud-federation#693](https://github.com/pangeo-data/pangeo-cloud-federation/issues/693)).

**Comment by droctothorpe** [comment]
Any updates on this? Any plans to tackle it in the near future? If not, I may try to pull this work into our next PI.

**Comment by TomAugspurger** [comment]
Nothing from me. It'd be great if you were able to take this up.

**Comment by jcrist** [comment]
There's some tricky aspects to this, especially from the k8s side. I'm hoping to have time to work on it sometime soon (rather busy with work right now), but no promises. I'd be happy if you wanted to give it a go, just know it's much less straightforward than the per-cluster limits implementation.
